flink-core/src/main/java/org/apache/flink/util/FileUtils.java

```diff
@@ -525,6 +540,102 @@ public static Path expandDirectory(Path file, Path targetDirectory) throws IOExc
 		return new Path(targetDirectory, rootDir);
 	}
 
+	/**
+	 * List the directory {@code directory} recursively and return the files that satisfies the {@code fileFilter}.
+	 *
+	 * @param directory the directory to be listed
+	 * @param fileFilter a file filter
+	 * @return a collection of {@code File}s
+	 *
+	 * @throws IOException if an I/O error occurs while listing the files in the given directory
+	 */
+	public static Collection<File> listFilesInPath(@Nonnull final File directory, @Nonnull final Predicate<File> fileFilter) throws IOException {
+		if (!Files.exists(directory.toPath())) {
+			throw new IllegalArgumentException(String.format("The directory %s dose not exist.", directory));
+		}
+		if (!Files.isDirectory(directory.toPath())) {
+			throw new IllegalArgumentException(String.format("The %s is not a directory.", directory));
+		}
+
+		final FilterFileVisitor filterFileVisitor = new FilterFileVisitor(fileFilter);
+
+		Files.walkFileTree(
+			directory.toPath(),
+			EnumSet.of(FileVisitOption.FOLLOW_LINKS),
+			Integer.MAX_VALUE,
+			filterFileVisitor
+		);
+		return filterFileVisitor.getFiles().isEmpty() ? Collections.emptyList() : filterFileVisitor.getFiles();
+	}
+
+	/**
+	 * Convert a collection of {@code File}s to a collection of relative path to the working dir.
+	 *
+	 * @param files a collection of files needed to be relatived
+	 * @return a collection of relative {@code File}s
+	 */
+	public static Collection<File> relativizeToWorkingDir(@Nonnull final Collection<File> files) {
+		final java.nio.file.Path workingDirPath = Paths.get(System.getProperty("user.dir"));
+
+		final List<File> relativeFiles = new LinkedList<>();
+
+		for (File file : files) {
+			if (file.isAbsolute()) {
+				relativeFiles.add(workingDirPath.relativize(file.toPath()).toFile());
+			} else {
+				relativeFiles.add(file);
+			}
+		}
+		return relativeFiles.isEmpty() ? Collections.emptyList() : relativeFiles;
+	}
+
+	/**
+	 * Convert a collection of relative {@code File}s to a collection of relative {@code URL}s.
+	 *
+	 * @param relativeFiles a collection of relative {@code File}s
+	 * @return a collection of relative URLs
+	 *
+	 * @throws MalformedURLException if error occurs while construct a url.
+	 */
+	public static Collection<URL> toRelativeURLs(@Nonnull final Collection<File> relativeFiles) throws MalformedURLException {
+		final List<URL> urls = new LinkedList<>();
+
+		for (File file : relativeFiles) {
+			urls.add(
+				new URL(
+					new URL(file.toURI().getScheme() + ":"),
+					file.toString()
```

Comment:
maybe `file.toString()` -> `file.getPath()` so we do not need to understand how `File.getString()` works.